libraries/core_libs/consensus/src/vote_manager/vote_manager.cpp

```diff
@@ -480,102 +403,34 @@ uint64_t VoteManager::roundDeterminedFromVotes(size_t two_t_plus_one) {
   return 0;
 }
 
-bool VoteManager::addRewardVote(const std::shared_ptr<Vote>& vote) {
-  const auto vote_hash = vote->getHash();
-  const auto voted_block_hash = vote->getBlockHash();
-  const auto vote_step = vote->getStep();
-  const auto vote_type = vote->getType();
-
-  if (vote_type != cert_vote_type) {
-    LOG(log_dg_) << "Drop vote " << vote_hash << ". The vote type " << vote_type << " is not cert vote type.";
-    return false;
-  }
+blk_hash_t VoteManager::getCurrentRewardsVotesBlock() const {
+  std::shared_lock lock(reward_votes_mutex_);
+  return reward_votes_pbft_block_hash_;
+}
 
-  if (vote_step != 3) {
-    LOG(log_dg_) << "Drop vote " << vote_hash << ". The vote step " << vote_step << " is not 3.";
+bool VoteManager::addRewardVote(const std::shared_ptr<Vote>& vote) {
+  std::unique_lock lock(reward_votes_mutex_);
+  if (vote->getBlockHash() != reward_votes_pbft_block_hash_) {
     return false;
   }
 
-  {
-    // This lock protects the both reward_votes_pbft_block_hash_ and reward_votes_
-    std::shared_lock lock(reward_votes_mutex_);
-
-    if (reward_votes_pbft_block_hash_ != voted_block_hash) {
-      // Should not happen, reward_votes_pbft_block_hash_ should always equal to last block hash in PBFT chain
-      LOG(log_wr_) << "The reward vote voted PBFT block hash " << voted_block_hash << " is different with "
-                   << reward_votes_pbft_block_hash_;
-      return false;
-    }
-
-    if (reward_votes_.contains(vote_hash)) {
-      // Should not happen, we have checked seen votes in VotePacketHandler::process
-      LOG(log_nf_) << "Reward vote " << vote_hash << " saved already";
+  // Limit reward_votes_ size by allowing only 1 unique vote per address. Otherwise attacker might send multiple valid
+  // reward votes with different round...
+  if (auto unique_vote_author = reward_votes_unique_authors_.insert({vote->getVoterAddr(), vote->getHash()});
+      !unique_vote_author.second) {
+    if (unique_vote_author.first->second != vote->getHash()) {
+      LOG(log_er_) << "Trying to add second reward vote: " << vote->getHash() << " for author: " << vote->getVoterAddr()
+                   << ", orig. vote: " << unique_vote_author.first->second;
       return false;
     }
   }
 
-  if (!verifyRewardVote(vote)) {
-    return false;
-  }
-
-  {
-    std::unique_lock lock(reward_votes_mutex_);
-    reward_votes_.insert({vote_hash, vote});
-  }
-
-  db_->saveLastBlockCertVote(vote);
```

Comment:
If you do not use it anywhere we should remove this function too

Reply:
yeah well it was used just here, but I have no idea why... Any possible clue why we would save block cert vote when adding reward vote ?

Reply:
I guess it was because it could be different cert vote we did not have saved before ?